ProcessTree.cpp

```diff
@@ -0,0 +1,530 @@
+}
+
+void ProcessTree::ApplyFlags(std::shared_ptr<ProcessTreeItem> process)
+{
+    unsigned int height = 0;
+    process->_flags = _filtersEngine->GetFlags(process, height);
+    if (process->_flags.none()) {
+        std::vector<struct Ancestor>::reverse_iterator rit = process->_ancestors.rbegin();
+        for (; rit != process->_ancestors.rend() && process->_flags.none(); ++rit) {
+            height++;
+            if (_processes.count(rit->pid) > 0) {
+                process->_flags = _filtersEngine->GetFlags(_processes[rit->pid], height);
+            }
+        }
+    }
+}
+
+
+void ProcessTree::PopulateTree()
+{ 
+    struct dirent *de;
+    int pid;
+
+    DIR *dr = opendir("/proc"); 
+  
+    if (dr == NULL)
+    { 
+        return;
+    } 
+  
+    while ((de = readdir(dr)) != NULL)  {
+        if (is_number(de->d_name)) {
+            pid = atoi(de->d_name);
+            auto process = ReadProcEntry(pid);
+            if (process != nullptr) {
+                _processes[pid] = process;
+            }
+        }
+    }
+
+    closedir(dr);     
+
+    for (auto p : _processes) {
+        auto process = p.second;
+        if (_processes.count(process->_ppid) > 0) {
+            _processes[process->_ppid]->_children.emplace_back(process->_pid);
+        }
+    }
+
+    for (auto p : _processes) {
+        std::shared_ptr<ProcessTreeItem> process, parent;
+        process = p.second;
+        if (_processes.count(process->_ppid) > 0) {
+            parent = _processes[process->_ppid];
+        } else {
+            parent = nullptr;
+        }
+        while (parent) {
+            process->_ancestors.insert(process->_ancestors.begin(), {parent->_pid, parent->_exe});
+            if (_processes.count(parent->_ppid) > 0) {
+                parent = _processes[parent->_ppid];
+            } else {
+                parent = nullptr;
+            }
+        }
+    }
+
+    for (auto p : _processes) {
+        ApplyFlags(p.second);
+    }
+} 
+
+std::string ProcessTree::ReadFirstLine(const std::string& file)
+{
+    std::ifstream f;
+    std::string line;
+
+    try {
+        f.open(file);
+        std::getline(f, line);
+        f.close();
+    } catch (...) {
+    }
+    return line;
+}
+
+std::string ProcessTree::ReadParam(const std::string& file, const std::string& param)
+{
+    std::ifstream f;
+    std::string line;
+    std::string value;
+
+    f.open(file);
+    while (f.good() && f.is_open() && !f.eof()) {
+        std::getline(f, line);
+        if (!line.compare(0, param.size() + 1, param + ":")) {
+            f.close();
+            value = line.substr(param.size() + 1);
+            size_t first = value.find_first_not_of(" \t");
+            if ( first == std::string::npos) {
+                return value;
+            } else {
+                return value.substr(first);
+            }
+        }
+    }
+    f.close();
+    return value;
+}
+
+std::shared_ptr<ProcessTreeItem> ProcessTree::ReadProcEntry(int pid)
```

Comment:
Why not use ProcessInfo::Open() ??

Reply:
ProcessInfo looked overly complex and I was concerned over the mixture of smart pointers and the 'new' construction. Instead I quickly built a simple couple of functions to achieve my limited requirements.